File: apps/backend/src/api/routes/projects.py
from fastapi import APIRouter, Depends, BackgroundTasks
from starlette import status
from starlette.responses import FileResponse

from api.dependencies import get_project_service
from models.project import ProjectResponse, CreateProjectRequest, UpdateProjectRequest
from services.project_service import ProjectService

PROJECT_PREFIX = "/projects"

router = APIRouter(prefix=PROJECT_PREFIX, tags=["projects"])


@router.post("/", response_model=ProjectResponse)
def create_project(
        request: CreateProjectRequest,
        project_service: ProjectService = Depends(get_project_service),
):
    return project_service.create_project(request)


@router.get("/", response_model=list[ProjectResponse])
def list_projects(
        projects_service: ProjectService = Depends(get_project_service),
):
    return projects_service.list_projects()


@router.get("/{project_id}", response_model=ProjectResponse)
def get_project(
        project_id: int,
        project_service: ProjectService = Depends(get_project_service),
):
    return project_service.get_project(project_id)


@router.patch("/{project_id}", response_model=ProjectResponse)
def update_project(
        project_id: int,
        request: UpdateProjectRequest,
        project_service: ProjectService = Depends(get_project_service),
):
    return project_service.update_project(project_id, request)

@router.delete("/{project_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_project(
        project_id: int,
        project_service: ProjectService = Depends(get_project_service),
):
    project_service.delete_project(project_id)

@router.get("/{project_id}/export")
def export_project(
        project_id: int,
        background_tasks: BackgroundTasks,
        project_service: ProjectService = Depends(get_project_service),
):
    export = project_service.create_export(project_id)
    background_tasks.add_task(project_service.cleanup_export, export.path)
    return FileResponse(
        path=export.path,
        status_code=status.HTTP_200_OK,
        media_type='application/zip',
        filename=export.name,
    )

@router.post("/import", response_model=ImportProjectReponse)
def import_project(
        project_service: ProjectService = Depends(get_project_service),
):
    return project_service.import_project(ImportProjectRequest())

File: packages/api/src/projects.ts
import {client} from "./client";
import {AgentResponse} from "./agents";
import {VideoResponse} from "./videos";

const api = client.extend((options) => ({
  prefix: `${options.prefix}/projects`,
}));

export interface ProjectResponse {
  id: number;
  label: string;
  agent: AgentResponse | null;
  videos: VideoResponse[];
  knowledge_id: number | null;
  idle_video: VideoResponse | null;
  enter_video: VideoResponse | null;
  exit_video: VideoResponse | null;
}

export interface CreateProjectRequest {
  label: string;
  agent_id: number;
  video_ids: number[];
  knowledge_id: number | null;
  idle_video_id: number | null;
  enter_video_id: number | null;
  exit_video_id: number | null;
}

export interface UpdateProjectRequest {
  label?: string | null;
  agent_id?: number | null;
  knowledge_id?: number | null;
  idle_video_id?: number | null;
  enter_video_id?: number | null;
  exit_video_id?: number | null;
}

export const projectsApi = {
  createProject: async (request: CreateProjectRequest): Promise<ProjectResponse> => {
    return await api.post("", {json: request}).json();
  },

  listProjects: async (): Promise<ProjectResponse[]> => {
    return await api.get("").json();
  },

  getProject: async (projectId: number): Promise<ProjectResponse> => {
    return await api.get(`${projectId}`).json();
  },

  updateProject: async (
    projectId: number,
    request: UpdateProjectRequest,
  ): Promise<ProjectResponse> => {
    return await api.patch(`${projectId}`, {json: request}).json();
  },

  deleteProject: async (projectId: number): Promise<void> => {
    await api.delete(`${projectId}`);
  },
};
